styx-scheduler-service/src/main/java/com/spotify/styx/state/QueuedStateManager.java

```diff
@@ -60,391 +58,227 @@
-  static final int SHUTDOWN_GRACE_PERIOD_SECONDS = 5;
-  static final long NO_EVENTS_PROCESSED = -1L;
+  private static final long NO_EVENTS_PROCESSED = -1L;
+
+  private static final Duration SHUTDOWN_GRACE_PERIOD = Duration.ofSeconds(5);
 
   private final Time time;
-  private final Executor outputHandlerExecutor;
+
+  private final ExecutorService outputHandlerExecutor;
   private final Storage storage;
   private final BiConsumer<SequenceEvent, RunState> eventConsumer;
   private final Executor eventConsumerExecutor;
+  private final OutputHandler outputHandler;
 
-  private final ConcurrentMap<WorkflowInstance, InstanceState> states = Maps.newConcurrentMap();
-
-  private final Thread dispatcherThread;
-  private final CountDownLatch closedLatch = new CountDownLatch(1);
-  private final Object signal = new Object();
-  private AtomicInteger activeEvents = new AtomicInteger(0);
   private volatile boolean running = true;
 
   public QueuedStateManager(
       Time time,
-      Executor outputHandlerExecutor,
+      ExecutorService outputHandlerExecutor,
       Storage storage,
       BiConsumer<SequenceEvent, RunState> eventConsumer,
-      Executor eventConsumerExecutor) {
+      Executor eventConsumerExecutor,
+      OutputHandler outputHandler) {
     this.time = Objects.requireNonNull(time);
     this.outputHandlerExecutor = Objects.requireNonNull(outputHandlerExecutor);
     this.storage = Objects.requireNonNull(storage);
     this.eventConsumer = Objects.requireNonNull(eventConsumer);
     this.eventConsumerExecutor = Objects.requireNonNull(eventConsumerExecutor);
-    this.dispatcherThread = new Thread(this::dispatch);
-    dispatcherThread.setName(DISPATCHER_THREAD_NAME);
-    dispatcherThread.start();
+    this.outputHandler = Objects.requireNonNull(outputHandler);
   }
 
   @Override
-  public void initialize(RunState runState) throws IsClosedException {
+  public CompletionStage<Void> trigger(WorkflowInstance workflowInstance, Trigger trigger) throws IsClosedException {
     ensureRunning();
 
-    final WorkflowInstance workflowInstance = runState.workflowInstance();
-    if (states.containsKey(workflowInstance)) {
-      throw new AlreadyInitializedException("RunState initialization called on active instance "
-                                            + workflowInstance.toKey());
-    }
-
-    final long counter;
+    final long nextCounter;
     try {
-      counter = storage.getLatestStoredCounter(workflowInstance).orElse(NO_EVENTS_PROCESSED);
+      final long counter = storage.getLatestStoredCounter(workflowInstance).orElse(NO_EVENTS_PROCESSED);
+      nextCounter = counter + 1;
     } catch (IOException e) {
-      throw Throwables.propagate(e);
+      return CompletableFutures.exceptionallyCompletedFuture(e);
     }
 
-    states.computeIfAbsent(workflowInstance, (wfi) -> new InstanceState(wfi, runState, counter + 1));
-  }
+    // TODO: optional retry on transaction conflict
```

Comment:
I don't think we want to retry on conflict. Edge cases might cause the same event to be processed twice. For example:
- `dequeue` event transition fails with conflict (and succeeds on another instance);
- instance is slow before retrying. In the meantime the workflow instance runs and gets back to QUEUED state;
- the previous `dequeue` event is retried and instance gets dequeued a second time (skipping the retry delay, etc.); 

EDIT: I wonder if this can still happen without retries on conflict. Two instances might try to `dequeue` at the same time but one gets slow before even starting the transaction while the other succeeds and a whole execution completes and the state goes back to QUEUED. Afterwards, the slow instance catches up and `dequeue` again (skipping retry delay, etc.).